src/agent/src/image_rpc.rs

```diff
@@ -0,0 +1,160 @@
+// Copyright (c) 2021 Alibaba Cloud
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+use std::fs;
+use std::process::{ExitStatus, Command};
+use std::sync::Arc;
+
+use anyhow::{anyhow, Result};
+use async_trait::async_trait;
+use protocols::image;
+use tokio::sync::Mutex;
+use ttrpc::{self, error::get_rpc_status as ttrpc_error};
+
+use crate::sandbox::Sandbox;
+//use crate::trace_rpc_call;
+use crate::rpc::CONTAINER_BASE;
+
+const SKOPEO_PATH: &str = "/usr/bin/skopeo";
+const UMOCI_PATH: &str = "/usr/local/bin/umoci";
+
+// Convenience macro to obtain the scope logger
+macro_rules! sl {
+    () => {
+        slog_scope::logger()
+    };
+}
+
+pub struct ImageService {
+    sandbox: Arc<Mutex<Sandbox>>,
+}
+
+impl ImageService {
+    pub fn new(sandbox: Arc<Mutex<Sandbox>>) -> Self {
+        Self { sandbox }
+    }
+
+    fn pull_image_from_registry(image: &str, cid: &str, source_creds: &Option<&str>) -> Result<()> {
+        let source_image = format!("{}{}", "docker://", image);
+
+        let manifest_path = format!("/tmp/{}/image_manifest", cid);
+        let target_path_manifest = format!("dir://{}", manifest_path);
+
+        // Define the target transport and path for the OCI image, without signature
+        let oci_path = format!("/tmp/{}/image_oci:latest", cid);
+        let target_path_oci = format!("oci://{}", oci_path);
+
+        fs::create_dir_all(&manifest_path)?;
+        fs::create_dir_all(&oci_path)?;
+
+        info!(sl!(), "Attempting to pull image {}...", &source_image);
+
+        let mut pull_command = Command::new(SKOPEO_PATH);
+        pull_command
+            .arg("--insecure-policy")
+            .arg("copy")
+            .arg(source_image)
+            .arg(&target_path_manifest);
+
+        if let Some(source_creds) = source_creds {
+            pull_command.arg("--src-creds").arg(source_creds);
+        }
+
+        let status: ExitStatus = pull_command.status()?;
+
+        if !status.success() {
+            return Err(anyhow!(format!("failed to pull image: {:?}", status)));
+        }
+
+        // Copy image from one local file-system to another
```

Comment:
Can you explain why we are having to do this copy?

Reply:
In my test, this step can actually be omitted. In one of the earlier versions(https://github.com/kata-containers/kata-containers/pull/2734) I submitted, it can directly:
```
skopeo copy docker://$IMAGE oci:$OCI_IMAGE
```
It can omit the manifest step.

Reply:
OCI format is needed. 
Eventually the spec and rootfs in the OCI format will be used, but manifest is not needed, it is only an intermediate format and was eventually removed at https://github.com/kata-containers/kata-containers/pull/2956/files#diff-a14bc6af50efd1f06bfe256b3d5d9a0b3be97955c120b9994e7a02b0996d2ff3R88L88.

Reply:
Two copies are currently used:
```
skopeo --insecure-policy copy docker://docker.io/library/busybox:latest dir:///tmp/<ID>/image_manifest
skopeo --insecure-policy copy dir:///tmp/<ID>/image_manifest oci:///tmp/<ID>image_oci:latest --remove-signatures
```
When I tested it, it seemed that once is fine:
```
skopeo copy  docker://docker.io/library/busybox:latest  oci:///tmp/<ID>image_oci:latest
```

Reply:
If things work with one single copy, then yes we need to do that.
@stevenhorsman @davidhay1969 Any idea why we're going through those 2 separate steps?

Reply:
Dave can probably confirm, but I think it it was 2 steps as the middle form was needed to run `skopeo standalone-verify` against to check the signature, but that shouldn't be required anymore with the plan in https://github.com/kata-containers/kata-containers/issues/2682

Reply:
From recollection, `umoci` cannot "unpack" an OCI image that contains a signature, hence we used `skopeo copy` on the second occasion to copy **from** the local manifest-formatted image **to** the OCI-formatted image, and remove the signatures

Reply:
@davidhay1969 @stevenhorsman 
Thanks for your comments.